src/System.IO.Pipelines/tests/PipeReaderCopyToAsyncTests.cs

```diff
@@ -265,11 +265,95 @@ public async Task ThrowingFromStreamDoesNotLeavePipeReaderInBrokenState()
             Assert.Equal(20, result.Buffer.Length);
             pipe.Reader.Complete();
         }
+
+        [Theory]
+        [InlineData(0)]
+        [InlineData(1)]
+        public async Task ThrowingFromStreamCallsAdvanceToWithStartOfLastReadResult(int throwAfterNWrites)
```

Comment:
Would this cover the case that I was seeing, where I was unable to write to the response stream cause the request stream was still being used?
Thanks

Reply:
If `Request.BodyPipe.CopyToAsync` was being called directly or indirectly on the previous request body, this could solve some problems. I don't know why the request stream being in a bad state would cause problems writing to the response stream though.

Reply:
My main issue is not really with CopyTo, its what happens when CopyTo leaves the request stream in a bad state. For some reason when the Request stream when written to throws this exception on the client "The stream does not support concurrent IO read or write operations.". 

I need the Request stream to be drained before CopyTo exists so that I can write a response to the response stream.

Thanks

Reply:
I don't think this fix will address https://github.com/dotnet/aspnetcore/issues/17723. That will likely require a change in FormPipeReader.ReadFormAsync() itself.

Kestrel can probably be hardened to call PipeReader.AdvanceTo() implicitly when the middleware does not, but that's a fairly large change that would also need to be made in the https://github.com/dotnet/aspnetcore repo. There's also a risk that sometimes AdvanceTo() isn't called because something isn't properly awaited, so implicitly calling AdvanceTo() is not without some risk. Ending request processing on the connection in this case might be the better solution.